### ISSUE TITLE
Go 1.18 BuildInfo support

### ISSUE THREAD

**Issue by abhinav** [bazelbuild/rules_go#3090]
### What version of rules_go are you using?

v0.30.0

### What version of gazelle are you using?

6bbfc47f1b0a27ee1efeddcc6671f3e4e03235dc

### What version of Bazel are you using?

```
Build label: 5.0.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Jan 19 14:08:54 2022 (1642601334)
Build timestamp: 1642601334
Build timestamp as int: 1642601334
```

### Does this issue reproduce with the latest releases of all the above?

Yes.

### What operating system and processor architecture are you using?

```
% uname -sm
Linux x86_64
```

and

```
% uname -sm
Darwin arm64
```

### What did you do?

Wrote a program that tries to access build information via [`runtime/debug.ReadBuildInfo`](https://pkg.go.dev/runtime/debug#ReadBuildInfo) and built it inside a Git repository.

```go
package main

import (
	"fmt"
	"runtime/debug"
)

func main() {
	info, ok := debug.ReadBuildInfo()
	fmt.Println(info, ok)
}
```

### What did you expect to see?

Information about the build environment.

### What did you see instead?

```
<nil> false
```

**Comment by sluongng** [comment]
You would need to use bazel stamp feature in combination with rules_go x_defs instead.

It does not make sense to enable this in rules_go by default as it affect build hermeticity and ruin bazel cache pretty badly.

**Comment by ar3s3ru** [comment]
> You would need to use bazel stamp feature in combination with rules_go x_defs instead.
> 
> It does not make sense to enable this in rules_go by default as it affect build hermeticity and ruin bazel cache pretty badly.

Hey @sluongng, would you mind providing more info on how to use this?

**Comment by sluongng** [comment]
Here you go https://github.com/bazelbuild/rules_go/blob/master/docs/go/core/defines_and_stamping.md#defines-and-stamping

**Comment by garymm** [comment]
Is it possible to have the default behavior of `bazel build --stamp` produce something that's usable with `debug.ReadBuildInfo()`?
If not possible to make it default, it'd be nice if this were documented.

**Comment by garymm** [comment]
Was poking around to see how `go build` does it, found this:
https://cs.github.com/golang/go/blob/245e95dfabd77f337373bf2d6bb47cd353ad8d74/src/cmd/go/internal/modload/build.go#L442

Still haven't looked at it long enough to figure out if it's feasible to replicate, or how, using rules_go.

**Comment by malt3** [comment]
Unsure if this should be a separate issue, but I think the solution cannot be stamping alone.

While I personally am happy to embed version info, timestamps and maybe a git commit via stamping, the `BuildInfo` contains a lot more than just this.
When using `go build`, this will also [place version info for all dependencies that end up in a binary][1] (think go.mod / go.sum) into  this buildinfo section.
This info can be extracted using `go version -m <GO_BINARY_FILE>` or `objdump -s -j .go.buildinfo <GO_BINARY_FILE>` and can be used to generate SBOMs just from the binary alone.
I feel like this is an important feature. It has also been around since go 1.13 so this part of the build info exists for far longer.

[1]: https://go.dev/doc/go1.13#go-command

**Comment by fmeum** [comment]
I most likely won't be able to work on this, but can promise timely reviews.

@aiuto Could Google be interested in funding this effort through the SIG to bootstrap SBOM support?

**Comment by malt3** [comment]
Thanks for the compilation of relevant info / steps. At this point in time I cannot offer to work on this feature due to time limitations and my unfamiliarity with writing bazel rules.

I would be super happy to see this implemented but I understand that resources are missing to work on this.

**Comment by udf2457** [comment]
I am seeing this issue (or something very similar) with:

`rules_go` 0.46.0
`gazelle` 0.35.0
`go_sdk` 1.22.1
`bazel` 7.0.2

**Comment by fmeum** [comment]
Just adding a working link to the relevant part of the Go codebase that explains how build info is sent to the linker: https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/work/exec.go;l=1527;drc=1ef6b2805ea1d78c082c7c55df39fc3d3942e4c9

**Comment by malt3** [comment]
And here is the code path that actually populates `BuildInfo`: https://cs.opensource.google/go/go/+/master:src/cmd/go/internal/load/pkg.go;l=2238;drc=1ef6b2805ea1d78c082c7c55df39fc3d3942e4c9;bpv=1;bpt=1

**Comment by apsaltis-ddog** [comment]
I've been looking into this and have a working prototype that can theoretically do everything but the deps.

The main challenge seems to be that much of the buildinfo data is tied to compilation, but it needs to be available then the binaries are linked, so it needs to be plumbed from the compilation of `main` to the link step.

Any thoughts on the approach?  It seems like we could perhaps have the compilation of the `main` package for `go_binary`s dump an output that does the right thing, or have a repeated builder link flag that specifies the buildinfo mapping.  I was initially leaning towards the latter.

**Comment by ddelnano** [comment]
@apsaltis-ddog are you still working on the prototype and would you be able to share a branch with the changes? We (Pixie) are keen to have this support as well.

**Comment by jsharpe** [comment]
@ddelnano https://github.com/jsharpe/rules_go/commit/71456d28e8b9fa40d524658fe4c3ff24c55ff51b is a version of this patched against rules_go 0.51 which is based on @apsaltis-ddog previous work. I'm working on rebasing against main to be able to get this merged upstream.

**Comment by ddelnano** [comment]
@jsharpe thanks for the update! Will keep my eye out on the linked PR

**Comment by darkrift** [comment]
@jsharpe any chance you can get this through the finish line ?

**Comment by jsharpe** [comment]
> @jsharpe any chance you can get this through the finish line ?

Yes, sorry this had dropped off my priority list; I'll likely get back to this early next week!

**Comment by HDYA** [comment]
This feature has been anticipated for quite some time, and I am eager to see your implementation. @jsharpe
